lib/upgrade/etcd/conversionv1v3/hostendpoint.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conversionv1v3
+
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apiv1 "github.com/projectcalico/libcalico-go/lib/apis/v1"
+	"github.com/projectcalico/libcalico-go/lib/apis/v1/unversioned"
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+	"fmt"
+	"github.com/projectcalico/libcalico-go/lib/numorstring"
+)
+
+type HostEndpoint struct{}
+
+// convertAPIToKVPair converts an API HostEndpoint structure to a KVPair containing a
+// backend HostEndpoint and HostEndpointKey.
+// This is part of the conversionHelper interface.
+func (h HostEndpoint) APIV1ToBackendV1(a unversioned.Resource) (*model.KVPair, error) {
+	ah := a.(apiv1.HostEndpoint)
+
+	var ipv4Addrs []cnet.IP
+	var ipv6Addrs []cnet.IP
+	for _, ip := range ah.Spec.ExpectedIPs {
+		if ip.Version() == 4 {
+			ipv4Addrs = append(ipv4Addrs, ip)
+		} else {
+			ipv6Addrs = append(ipv6Addrs, ip)
+		}
+	}
+
+	var ports []model.EndpointPort
+	for _, port := range ah.Spec.Ports {
+		ports = append(ports, model.EndpointPort{
+			Name:     port.Name, // name
+			Protocol: port.Protocol, // protocol
+			Port:     port.Port, //gt=0
+		})
+	}
+
+	d := model.KVPair{
+		Key: model.HostEndpointKey{
+			Hostname:   ah.Metadata.Node, // name
+			EndpointID: ah.Metadata.Name, // namespacedName
+		},
+		Value: &model.HostEndpoint{
+			Labels: ah.Metadata.Labels,
+
+			Name:              ah.Spec.InterfaceName,
+			ProfileIDs:        ah.Spec.Profiles,
+			ExpectedIPv4Addrs: ipv4Addrs,
+			ExpectedIPv6Addrs: ipv6Addrs,
+			Ports:             ports,
+		},
+	}
+
```

Comment:
also, can you add a log line at the end of both conversions so we know what it converted to and from